src/ray/core_worker/core_worker.cc

```diff
@@ -281,13 +283,47 @@ void CoreWorker::ReportActiveObjectIDs() {
   heartbeat_timer_.async_wait(boost::bind(&CoreWorker::ReportActiveObjectIDs, this));
 }
 
-void CoreWorker::PromoteObjectToPlasma(const ObjectID &object_id) {
+bool CoreWorker::SerializeObjectId(const ObjectID &object_id, TaskID *owner_id,
+                                   rpc::Address *owner_address) {
   RAY_CHECK(object_id.IsDirectCallType());
   auto value = memory_store_->GetOrPromoteToPlasma(object_id);
   if (value != nullptr) {
     RAY_CHECK_OK(
         plasma_store_provider_->Put(*value, object_id.WithPlasmaTransportType()));
   }
+
+  auto has_owner = reference_counter_->GetOwner(object_id, owner_id, owner_address);
+  if (!has_owner) {
+    // TODO(swang): Once ref counting is fully implemented, this should be an
+    // assertion.
+    RAY_LOG(ERROR) << "Tried to serialize object ID " << object_id
+                   << ", but we don't know its owner. Calling ray.get on this object may "
+                      "throw an error if the object takes too long to create.";
+  }
+  return has_owner;
+}
+
+void CoreWorker::DeserializeObjectId(const ObjectID &object_id, const TaskID &owner_id,
+                                     const rpc::Address &owner_address) {
+  // Add the object's owner to the local metadata in case it gets serialized
+  // again.
+  reference_counter_->AddBorrowedObject(object_id, owner_id, owner_address);
+
+  // TODO: Set the object as in plasma immediately if owner is nil.
+  if (owner_id.IsNil()) {
+    // No owner found so we cannot find out the object's status. Store an
+    // IsInPlasmaError immediately so that the caller will try to fetch it via
+    // plasma.
+    // NOTE(swang): For objects that do not have owners, this may cause
+    // ray.gets to timeout with UnreconstructableError if the task that is
+    // supposed to create the object is too slow.
+    RAY_CHECK_OK(
+        memory_store_->Put(RayObject(rpc::ErrorType::OBJECT_IN_PLASMA), object_id));
```

Comment:
Don't we enforce the invariant that serialized objects are always promoted to plasma, and hence we will always be fetching it from plasma anyways?

Reply:
Ok, I think I see what you're doing. It strikes me as a bit of a hack though -- why not let instead let plasma fetches block indefinitely in the plasma provider, and interrupt the fetch as needed?

Reply:
I.e., preserving the invariants of "direct call ids point to local objects", and "plasma ids point to plasma objects" seems desirable.

Reply:
Agree with this. It would be cleaner for memory store to only contain direct call objects